KAFKA-14412: Better Rocks column family management

nicktelford opened the pull request:
When opening RocksDB, we were checking for an error in `RocksDBTimestampedStore` to detect if the `keyValueWithTimestamp` CF is missing.

The `openRocksDB` method now supports any number of column families, not just the extra one used by `RocksDBTimestampedStore`. We now check for the existing column families _before_ opening the database, which allows us to create any missing column families.

Supporting automatic creation of any number of missing column families is a pre-requisite for KIP-892: Transactional StateStores.

nicktelford committed:
KAFKA-14412: Better Rocks column family management

When opening RocksDB, we were checking for an error in
`RocksDBTimestampedStore` to detect if the `keyValueWithTimestamp` CF is
missing.

The `openRocksDB` method now supports any number of column families, not
just the extra one used by `RocksDBTimestampedStore`. We now check for
the existing column families _before_ opening the database, which allows
us to create any missing column families.

Supporting automatic creation of any number of missing column families
is a pre-requisite for KIP-892: Transactional StateStores.

nicktelford commented:
This is part of KIP-892, and has been broken out into a separate PR to reduce the review burden on the main KIP-892 implementation, since it can be merged independently.

Note: KIP-892 requires this, as it introduces an additional column family for offset management to both `RocksDBStore` and `RocksDBTimestampedStore`, and the previous approach (checking for an error on-open) was not scalable to more than 1 non-default column family.

nicktelford commented:
@cadonna

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -278,13 +280,55 @@ private void addValueProvidersToMetricsRecorder() {
+            final List<byte[]> allExisting = RocksDB.listColumnFamilies(options, absolutePath);
+
+            final List<ColumnFamilyDescriptor> existingDescriptors = allDescriptors.stream()
+                    .filter(descriptor -> descriptor == defaultColumnFamilyDescriptor || allExisting.stream().anyMatch(existing -> Arrays.equals(existing, descriptor.getName())))
```
Wouldn't `allExisting.stream().anyMatch(existing -> Arrays.equals(existing, descriptor.getName()))` be enough?
Should the addition of `descriptor == defaultColumnFamilyDescriptor` be a performance improvement?

nicktelford replied:
This is necessary because in the case that the database is being opened for the first time, the default column family won't (yet) exist, and so `listColumnFamilies` returns an empty `List`.

Since the default column family will be created when we call `open` later, we don't need to explicitly create it, so we can count it as an `existingDescriptor`, even if it doesn't yet exist.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -278,13 +280,55 @@ private void addValueProvidersToMetricsRecorder() {
+                } else if (created < toCreate.size() && (toCreate.get(created) == allDescriptors.get(existing + created))) {
+                    columnFamilies.add(createdColumnFamilies.get(created));
+                    created++;
+                }
```
Maybe add an `else`-branch that throws an `IllegalStateException` to catch bugs.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -278,13 +280,55 @@ private void addValueProvidersToMetricsRecorder() {
+            db = RocksDB.open(dbOptions, absolutePath, existingDescriptors, existingColumnFamilies);
+            final List<ColumnFamilyHandle> createdColumnFamilies = db.createColumnFamilies(toCreate);
+
+            // match up the existing and created ColumnFamilyHandles with the existing/created ColumnFamilyDescriptors
+            // so that the order of the resultant List matches the order of the openRocksDB arguments
+            final List<ColumnFamilyHandle> columnFamilies = new ArrayList<>(allDescriptors.size());
+            int existing = 0;
+            int created = 0;
+            while (existing + created < allDescriptors.size()) {
+                if (existing < existingDescriptors.size() && (existingDescriptors.get(existing) == allDescriptors.get(existing + created))) {
+                    columnFamilies.add(existingColumnFamilies.get(existing));
+                    existing++;
+                } else if (created < toCreate.size() && (toCreate.get(created) == allDescriptors.get(existing + created))) {
+                    columnFamilies.add(createdColumnFamilies.get(created));
+                    created++;
+                }
+            }
```
Could you please extract this code to method with a meaningful name such that you do not need the inline comment? 
If that does not work well, then comment on the extracted method.

nicktelford replied:
Done. I've also refactored the method a bit to avoid depending on the `existingDescriptor` and `toCreate` lists, since that would have required the method to have 5 arguments; it now only has 3.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -278,13 +280,55 @@ private void addValueProvidersToMetricsRecorder() {
+    /**
+     * Open RocksDB while automatically creating any requested column families that don't yet exist.
+     */
+    List<ColumnFamilyHandle> openRocksDB(final DBOptions dbOptions,
```
Shouldn't that be `protected` since `RocksDBTimestampedStore` extends `RocksDBStore`?

cadonna reviewed (commented):
Thanks for the PR, @nicktelford!

Here my feedback!

nicktelford committed:
Add IllegalStateException branch

If our column family handles and descriptors for some reason don't
match-up, we should throw an exception, otherwise we would end up in an
infinite loop, since neither of our loop counters get incremented.

nicktelford committed:
Make openRocksDb method protected

This is only used by `RocksDBStore` and `RocksDBTimestampedStore`, and
never used by tests, so it should be protected from incorrect use.

nicktelford committed:
Extract and refactor mergeColumnFamilyHandleLists

The method has been simplified to no longer depend on the
`existingDescriptors` and `toCreate`, which would have made this method
much more cumbersome to understand.

We now directly compare the name of the column family handle with that
of its descriptor.

nicktelford commented:
@mjsax @ableegoldman @lucasbru @wcarlson5 @bbejeck @vvcephei @guozhangwang 
I was recommended to tag you all in these Kafka Streams PRs.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java:
```diff
@@ -278,18 +280,74 @@ private void addValueProvidersToMetricsRecorder() {
+
+            final List<ColumnFamilyDescriptor> existingDescriptors = allDescriptors.stream()
+                    .filter(descriptor -> descriptor == defaultColumnFamilyDescriptor || allExisting.stream().anyMatch(existing -> Arrays.equals(existing, descriptor.getName())))
+                    .collect(Collectors.toList());
```
nit:
Just a proposal for better readability:
```suggestion
            final List<ColumnFamilyDescriptor> existingDescriptors = new LinkedList<>();
            existingDescriptors.add(defaultColumnFamilyDescriptor);
            existingDescriptors.addAll(allDescriptors.stream()
                .filter(descriptor -> allExisting.stream().anyMatch(existing -> Arrays.equals(existing, descriptor.getName())))
                    .collect(Collectors.toList()
                );
```

cadonna reviewed (commented):
@nicktelford Thanks for the updates!

Do we have unit tests in place that test all this logic?

nicktelford committed:
Improve readability of openRocksDB

Adding the `defaultColumnFamilyDescriptor` explicitly, instead of filtering on it, makes it more obvious what we're doing.

Co-authored-by: Bruno Cadonna <cadonna@apache.org>

nicktelford commented:
@cadonna 
> Do we have unit tests in place that test all this logic?

We don't have any tests dedicated to this, but `RocksDBStoreTest` and `RocksDBTimestampedStoreTest` implicitly test both the case that we only have the default CF, and the case that we have an extra CF (for the timestamped values). These tests fail spectacularly when any of this logic has a bug in it.

Also, most of the integration tests break pretty badly when any of this is wrong.

Are there any additional tests you'd like me to add?

cadonna commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -64,45 +62,27 @@ public RocksDBTimestampedStore(final String name,
+                new ColumnFamilyDescriptor(RocksDB.DEFAULT_COLUMN_FAMILY, columnFamilyOptions),
+                new ColumnFamilyDescriptor("keyValueWithTimestamp".getBytes(StandardCharsets.UTF_8), columnFamilyOptions)
+        );
+        final ColumnFamilyHandle noTimestampCF = columnFamilies.get(0);
+        final ColumnFamilyHandle withTimestampCF = columnFamilies.get(1);
```
nit: Could you please spell out `ColumnFamily`?

nicktelford replied:
Done.

cadonna reviewed (commented):
@nicktelford Thank you for the clarification!

There is a compilation error that let the builds fail.

When you fix this could you please also consider the nit I added?

Regarding tests, it is always good if existing tests still run. I was more referring if you think we need to test this logic with existing column families that are not part of the two we currently use. I think we do not need them because, we do not really consider that use case.

nicktelford committed:
Fix compile errors and CF naming

nicktelford commented:
Sorry about the compile error. It should now build.

Regarding tests: agreed. When KIP-892 lands, it adds an extra CF, but at that point RocksDBStoreTest and RocksDBTimestampedStoreTest will implicitly test that too, so I don't think additional tests are required.

cadonna reviewed (approved):
Thanks for the updates, @nicktelford !

LGTM!

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimestampedStore.java:
```diff
@@ -64,31 +62,14 @@ public RocksDBTimestampedStore(final String name,
+        final List<ColumnFamilyHandle> columnFamilies = openRocksDB(
+                dbOptions,
+                new ColumnFamilyDescriptor(RocksDB.DEFAULT_COLUMN_FAMILY, columnFamilyOptions),
+                new ColumnFamilyDescriptor("keyValueWithTimestamp".getBytes(StandardCharsets.UTF_8), columnFamilyOptions)
```
We should define a constant for `keyValueWithTimestamp` instead of hard-coding the column family name like this. Especially since we'll be adding additional column families. We should have them all defined in one central place

nicktelford replied:
FWIW, this is how it currently is in trunk, but I agree and I can add a constant in this changeset.

nicktelford replied:
Any suggestions on the "one central place" to define them? This one only makes sense in RocksDBTimestampedStore, so defining them all in RocksDBStore may not make sense

cadonna replied:
IMO, it would fine to have a public constant for the name of the column family in `RocksDBTimestampedStore` since this column family only applies to this class and its tests. For the other more generic column family that you are going to add, I would say `RocksDBStore` is the right place.

nicktelford replied:
Done.

nicktelford committed:
Extract column family name to constant

For improved readability, and slightly better performance, we should
extract the name of the column family for timestamped values to a
constant.

When we later add a column family for offsets, a constant for that will
be defined in `RocksDBStore` with a similar name.

nicktelford committed:
Fix bug introduced in refactoring

Since we add the default column family explicitly, we only want to add
the `extraDescriptors`, to prevent adding the default column family
twice.

nicktelford commented:
I've rebased against trunk so this should be good to go.
